agent/eni/watcher_linux.go

```diff
@@ -42,50 +42,43 @@ type UdevWatcher struct {
 	netlinkClient        netlinkwrapper.NetLink
 	udevMonitor          udevwrapper.Udev
 	events               chan *udev.UEvent
-	state                eniStateManager.StateManager
+	agentState           dockerstate.TaskEngineState
+	eniChangeEvent       chan<- statechange.Event
 }
 
 // New is used to return an instance of the UdevWatcher struct
 func New(ctx context.Context, udevwrap udevwrapper.Udev,
 	state dockerstate.TaskEngineState, stateChangeEvents chan<- statechange.Event) *UdevWatcher {
-	return _new(ctx, netlinkwrapper.New(), udevwrap,
-		eniStateManager.New(state, stateChangeEvents))
+	return _new(ctx, netlinkwrapper.New(), udevwrap, state, stateChangeEvents)
 }
 
 // _new is used to nest the return of the UdevWatcher struct
-func _new(ctx context.Context, nlWrap netlinkwrapper.NetLink, udevWrap udevwrapper.Udev, eniState eniStateManager.StateManager) *UdevWatcher {
+func _new(ctx context.Context,
+	nlWrap netlinkwrapper.NetLink,
+	udevWrap udevwrapper.Udev,
+	state dockerstate.TaskEngineState,
+	stateChangeEvents chan<- statechange.Event) *UdevWatcher {
+
 	derivedContext, cancel := context.WithCancel(ctx)
 	return &UdevWatcher{
-		ctx:           derivedContext,
-		cancel:        cancel,
-		netlinkClient: nlWrap,
-		udevMonitor:   udevWrap,
-		events:        make(chan *udev.UEvent),
-		state:         eniState,
+		ctx:            derivedContext,
+		cancel:         cancel,
+		netlinkClient:  nlWrap,
+		udevMonitor:    udevWrap,
+		events:         make(chan *udev.UEvent),
+		agentState:     state,
+		eniChangeEvent: stateChangeEvents,
 	}
 }
 
 // Init initializes a new ENI Watcher
 func (udevWatcher *UdevWatcher) Init() error {
-	links, err := udevWatcher.netlinkClient.LinkList()
-	if err != nil {
-		return errors.Wrapf(err, "udev watcher init: error retrieving network interfaces")
-	}
-
-	// Return on empty list
-	if len(links) == 0 {
-		return errors.New("udev watcher init: no network interfaces discovered for initialization")
-	}
-
-	// Initialize eni state manager with a list of network interface
-	// devices discovered.
-	go udevWatcher.state.Init(links)
-	return nil
+	return udevWatcher.reconcileOnce()
```

Comment:
The only path that leads to an error here is the `LinkList`, which can fail because of failed syscalls and I think it's fine to not start the Agent if we are running into that. Moreover, this is sticking to the previous behavior anyway.